src/GafferImage/Catalogue.cpp

```diff
@@ -417,6 +417,17 @@ class Catalogue::InternalImage : public ImageNode
 					return nullptr;
 				}
 
+				// Its possible for a Catalogue to have been removed from its script
+				// and still receive an image. If it will attempt to save that image
+				// to a file which needed the script context to resolve properly,
+				// the saving will eventually error, so we return null instead.
+				// Its likely this only occurs while the node is in the process
+				// of being deleted (perhaps inside python's garbage collector).
+				if( !client->ancestor<ScriptNode>() && Context::hasSubstitutions( fileName ) )
```

Comment:
I think this would be better done in `generateFileName()` itself, so the fix applies everywhere it might be used, and so AsynchronousSaver doesn't need to know anything at all about how the filename is generated. `generateFileName()` can just return "" in this case, and then we'll hit the existing early-out above.